### DISCUSSION TITLE
Possible brewRec improvement

### DISCUSSION THREAD

**Discussion by coreywoodfield** [yakivy/jam#11] [Ideas]
I don't know how feasible this would be. But I think it'd be really cool. `brewRec` is super convenient, but I hesitate to use it for anything serious because of how easy it would be to misuse. If you want to create at most one instance of anything, you need to manually brew anything that is used by multiple things.

Example: with the following class definitions,
```scala
class A()
class B(a: A)
class C(a: A)
```
the following usage of the macro
```scala
trait Module {
  lazy val b: B = jam.brewRec[B]
  lazy val c: C = jam.brewRec[C]
}
```
currently creates something like
```scala
trait Module {
  lazy val b: B = new B(new A())
  lazy val c: C = new C(new A())
}
```
It'd be really nice if it could instead create something like
```scala
trait Module {
  lazy val _jam_brewRec_a: A = new A()
  lazy val b: B = new B(_jam_brewRec_a)
  lazy val c: C = new C(_jam_brewRec_a)
}
```

Would something like this be feasible? I'm guessing it wouldn't be possible without changing how the macro is invoked. Maybe `brewRec` could take a list of types to instantiate, and it could create fields for each of them and all their dependencies? Just an idea

**Comment by yakivy** [comment]
Hi @coreywoodfield.
Yes, in Scala 3 it's not possible to modify a scope outside macro application. 
Could you explain your problem in a more production-like basis? I usually have only several classes that require to be a singleton (http client, db connection pool, etc...) and in 99,9% they have unbrewable constructor, so even if you forget to create them as singletons compiler will throw an error. I tried to solve that 0.01% problem, but was not able to come up with one-fit-all solution, last idea was about limiting auto brewing by package name.

**Comment by yakivy** [comment]
Hi @coreywoodfield, I've finally got some free time to implement the idea I was talking about.
With jam `0.2.0` you can provide brewing configuration like:
```
object myjam extends jam.JamDsl {
    implicit inline def myJamConfig: JamConfig = JamConfig(brewRecRegex = "com\\.my\\..*")
}
```
and then calls for recursive brewing `myjam.brewRec[A]` will create automatically only classes from `com.my` package.
More info you can find here: https://github.com/yakivy/jam#macro-configuration
